templates/commands/render_action.go

```diff
@@ -37,65 +37,85 @@ import (
 // to be written.
 type walkAndModifyVisitor func([]byte) ([]byte, error)
 
-// Recursively traverses the directory or file scratchDir/relPath, calling the
-// given visitor for each file. If the visitor returns modified file contents
-// for a given file, that file will be overwritten with the new contents.
-func walkAndModify(ctx context.Context, pos *model.ConfigPos, rfs renderFS, scratchDir, relPath string, v walkAndModifyVisitor) error {
+// For each given path, recursively traverses the directory or file
+// scratchDir/relPath, calling the given visitor for each file. If the visitor
+// returns modified file contents for a given file, that file will be
+// overwritten with the new contents.
+//
+// A file will only be touched once per call, even if multiple paths include
+// it.
+func walkAndModify(ctx context.Context, rfs renderFS, scratchDir string, relPaths []model.String, v walkAndModifyVisitor) error {
 	logger := logging.FromContext(ctx).Named("walkAndModify")
+	seen := map[string]struct{}{}
 
-	relPath, err := safeRelPath(pos, relPath)
-	if err != nil {
-		return err
-	}
-	walkFrom := filepath.Join(scratchDir, relPath)
-	if _, err := rfs.Stat(walkFrom); err != nil {
-		if os.IsNotExist(err) {
-			return model.ErrWithPos(pos, `path %q doesn't exist in the scratch directory, did you forget to "include" it first?"`, relPath) //nolint:wrapcheck
-		}
-		return model.ErrWithPos(pos, "Stat(): %w", err) //nolint:wrapcheck
-	}
-
-	return filepath.WalkDir(walkFrom, func(path string, d fs.DirEntry, err error) error { //nolint:wrapcheck
+	for _, relPathPos := range relPaths {
+		relPathPos := relPathPos
```

Comment:
Is this necessary? I don't see that relPathPos appears in any closures or escapes a single loop iteration in any way. But I could be missing something.

Reply:
No, I'll remove.